### ISSUE TITLE
Suspected memory leak in KafkaThrottledLatestProcessedCommit

### ISSUE THREAD

**Issue by tilakchowdary** [smallrye/smallrye-reactive-messaging#2623]
I have observed that our heap slowly increasing with KafkaThrottledLatestProcessedCommit

Here is the heap map 
<img width="1623" alt="Screenshot 2024-05-17 at 1 54 35 PM" src="https://github.com/smallrye/smallrye-reactive-messaging/assets/25810378/ab4a9071-ab58-4235-bf66-1e9bfeddd5d3">
Any inputs are greatly appreciated

**Comment by cescoffier** [comment]
We would need to see the application. There is a poison pill mechanism that makes sure we do not accumulate too much.

So make sure your messages are acknowledged.

**Comment by ozangunalp** [comment]
Yes, probably your messages are not acknowledged properly, so not cleared from tracked messages.

**Comment by tilakchowdary** [comment]
@cescoffier @ozangunalp Thanks for the suggestion. As I understand if it was due to messages are not being acknowledged properly it would result in TooManyMessagesWithoutAckException. However, In our case we don't see this issue and we see uninterrupted records consumption. From heap analysis only the processedOffsets field was leaking.
I suspect concurrency usage of KafkaThrottledLatestProcessedCommit.
Methods processed(long offset) and clearLesserSequentiallyProcessedOffsetsAndReturnLargestOffset() - Calling these concurrently by multiple threads resulting inconsistent view of processedOffsets.

**Comment by cescoffier** [comment]
Aren't these methods getting called from event loops ?

**Comment by ozangunalp** [comment]
If you've specific logs linked to this issue it can be handy. Otherwise, I don't see why this might happen.

**Comment by cescoffier** [comment]
@ozangunalp maybe if `concurrency > 2`, so multiple event loops. But still, should not lead to a leak.

**Comment by cescoffier** [comment]
Yeah, that's what I just saw.

**Comment by tilakchowdary** [comment]
I think issue is that with IncomingKafkaRecordBatch only [latestOffsetRecords](https://github.com/smallrye/smallrye-reactive-messaging/blob/main/smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/KafkaSource.java#L360C47-L360C69) of that batch are stored  in OffsetStore.receivedOffsets and subsequently removed from processedOffsets. But processedOffsets will have all Offsets  not just the latestOffsetRecords.

**Comment by ozangunalp** [comment]
I'll verify batch consuming for leaks. Are you acking individual messages and not the batch as a whole?

**Comment by ozangunalp** [comment]
I didn't have time to reproduce this but acking individual messages in batch mode should be the reason for Offset tracking not getting cleaned
